File: amber/typescript/src/incoming.test.ts
import { ProvenanceContext } from "./context.js";
import {
  MAX_INCOMING_JSON_BYTES,
  inspectIncomingJSON,
  withIncomingJSON,
} from "./incoming.js";
import { Provenance } from "./provenance.js";

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

const root = Provenance.start();
const input = JSON.stringify(root);
const accepted = inspectIncomingJSON(input, "reject");
assert(accepted.present && accepted.provenance?.execution_id === root.execution_id, "valid input was not accepted");

const ignored = inspectIncomingJSON('{"version":1}', "ignore");
assert(!ignored.present && ignored.provenance === undefined, "invalid ignored input must be absent");

let rejected = false;
try {
  inspectIncomingJSON('{"version":1}', "reject");
} catch {
  rejected = true;
}
assert(rejected, "invalid rejected input must throw");

const oversized = "x".repeat(MAX_INCOMING_JSON_BYTES + 1);
assert(!inspectIncomingJSON(oversized, "ignore").present, "oversized ignored input must be absent");
rejected = false;
try {
  inspectIncomingJSON(oversized, "reject");
} catch {
  rejected = true;
}
assert(rejected, "oversized rejected input must throw");

const empty = ProvenanceContext.empty();
const installed = withIncomingJSON(empty, input, "reject");
assert(installed.present && installed.context.provenance?.execution_id === root.execution_id, "incoming context was not installed");
const unchanged = withIncomingJSON(installed.context, "", "reject");
assert(!unchanged.present && unchanged.context === installed.context, "absent input must preserve context");

console.log("TypeScript incoming context tests passed");

File: amber/go/incoming_test.go
package amber

import (
	"context"
	"encoding/json"
	"testing"
)

func TestIncomingInspectionPolicies(t *testing.T) {
	root, err := Start()
	if err != nil {
		t.Fatal(err)
	}
	data, err := json.Marshal(root)
	if err != nil {
		t.Fatal(err)
	}

	accepted, present, err := InspectIncomingJSON(data, IncomingReject)
	if err != nil || !present || accepted.ExecutionID() != root.ExecutionID() {
		t.Fatalf("valid incoming value was not accepted: present=%v err=%v", present, err)
	}
	if _, present, err := InspectIncomingJSON([]byte(`{"version":1}`), IncomingIgnore); err != nil || present {
		t.Fatalf("invalid ignored input should be absent: present=%v err=%v", present, err)
	}
	if _, present, err := InspectIncomingJSON([]byte(`{"version":1}`), IncomingReject); err == nil || present {
		t.Fatalf("invalid rejected input should return an error: present=%v err=%v", present, err)
	}
	oversized := make([]byte, MaxIncomingJSONBytes+1)
	if _, present, err := InspectIncomingJSON(oversized, IncomingIgnore); err != nil || present {
		t.Fatalf("oversized ignored input should be absent: present=%v err=%v", present, err)
	}
	if _, present, err := InspectIncomingJSON(oversized, IncomingReject); err == nil || present {
		t.Fatalf("oversized rejected input should return an error: present=%v err=%v", present, err)
	}
}

func TestWithIncomingJSONDoesNotOverwriteOnAbsentInput(t *testing.T) {
	root, err := Start()
	if err != nil {
		t.Fatal(err)
	}
	data, err := json.Marshal(root)
	if err != nil {
		t.Fatal(err)
	}
	base := context.Background()
	installed, present, err := WithIncomingJSON(base, data, IncomingReject)
	if err != nil || !present {
		t.Fatalf("valid incoming context was not installed: present=%v err=%v", present, err)
	}
	unchanged, present, err := WithIncomingJSON(installed, nil, IncomingReject)
	if err != nil || present || unchanged != installed {
		t.Fatalf("absent input must preserve the original context: present=%v err=%v", present, err)
	}
	current, ok := ProvenanceFromContext(unchanged)
	if !ok || current.ExecutionID() != root.ExecutionID() {
		t.Fatal("existing context was overwritten or lost")
	}
}
